### ISSUE TITLE
Mutators cause false positives in the nullness checker

### ISSUE THREAD

**Issue by sxlijin** [typetools/checker-framework#4689]
The nullness checker appears to be unable to determine that mutators do not modify a given variable. A minimal example that can be used to prove this:

```
import org.checkerframework.checker.nullness.qual.Nullable;
import java.util.ArrayList;

class Foo {
    static class Bar {
        Object a = new Object();
        void mutate() { a = new Object(); }
    }
    
    Bar b = new Bar();
    @Nullable Object o = null;
    
    void foo() {
        if (o == null) return;
        b.mutate();
        o.hashCode();
    }
}
```

The real-world example where this came up was something like:

```
List<Node> nodes = new ArrayList<>();
InterestingType a = null;
void loop() {
  ...
  if (a == null) {
    a = nonNullThing;
    nodes.add(buildNode());
    a.doThing();
  }
  ...
}
```

**Comment by mernst** [comment]
> The nullness checker appears to be unable to determine that mutators do not modify a given variable.

While checking one method, the Nullness Checker does not do inference of side effects for other methods.
Rather, the Nullness Checker verifies programmer-written annotations.

If you use a fact when reasoning about your code, you need to [express that fact as an annotation](https://checkerframework.org/manual/#handling-warnings).

The manual section ["Side effects"](https://checkerframework.org/manual/#type-refinement-side-effects) describes several ways to express that a mutator does not set a variable to null, including:
 * [`@SideEffectFree`](https://checkerframework.org/api/org/checkerframework/dataflow/qual/SideEffectFree.html)
 * [`@MonotonicNonNull`](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/MonotonicNonNull.html)
 * [`@EnsuresNonNull`](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/EnsuresNonNull.html)

It seems that `@MonotonicNonNull` might be appropriate for your real-world example.
I'm closing this issue, but let us know if the manual is not clear.

**Comment by sxlijin** [comment]
That is... very surprising.

I remember that it took me a _long_ time to track this down. Is there any way to perhaps propagate this information up to the resulting error, so that users at least know where to look?

**Comment by mernst** [comment]
Could you clarify what is the very surprising part?

Issue https://github.com/typetools/checker-framework/issues/982 is a proposal to give users more information about the fact that side-effecting methods may set fields.  Would that have helped you?
